plugins/jobs/src/js/JobStopRun.js

```diff
@@ -0,0 +1,70 @@
+import * as React from "react";
+import { componentFromStream } from "data-service";
+import { stopJobRun } from "#SRC/js/events/MetronomeClient";
+
+import { Subject } from "rxjs/Subject";
+import "rxjs/add/operator/map";
+import "rxjs/add/operator/switchMap";
+import "rxjs/add/operator/do";
+import "rxjs/add/operator/combineLatest";
+import "rxjs/add/operator/startWith";
+import "rxjs/add/operator/catch";
+
+import JobStopRunModal from "./components/JobStopRunModal";
+
+function executeStop({ jobID, jobRun, onSuccess }) {
+  return stopJobRun(jobID, jobRun)
+    .map(_ => ({ done: true }))
+    .do(_ => onSuccess())
+    .startWith({ done: false });
+}
+
+function stopOperation() {
+  const stopSubject$ = new Subject();
+  const stop$ = stopSubject$.switchMap(executeStop).catch(error => {
+    return stop$.startWith({
+      errorMsg: error.response.message,
+      done: true
+    });
+  });
+
+  return {
+    stop$,
+    stopHandler: (jobID, jobRun, onSuccess) => {
+      stopSubject$.next({ jobID, jobRun, onSuccess });
+    }
+  };
+}
+
+export function stopSingleJobRun(stopHandler, jobId, jobRuns, onSuccess) {
+  // TODO DCOS-8763 introduce support for multiple job run ID
+  if (jobRuns.length === 1) {
+    stopHandler(jobId, jobRuns[0], onSuccess);
+  }
+}
+
+export const JobStopRun = componentFromStream(props$ => {
+  const { stop$, stopHandler } = stopOperation();
+  const stopEmit$ = stop$.startWith({ done: null });
+
+  return props$
+    .combineLatest(stopEmit$, (props, stopOperation) => {
+      return { ...props, ...stopOperation };
```

Comment:
this is _really_ hard to reason about, what is the content of `props` and especially `stopOperation` here? 😢 please add some kind of types for both somewhere or dont use spread operator.

Reply:
Added prop types, remove the splat of the stop stream, renamed also to make it more clear. I still splat props, but now there is no ambiguity on which stream the data comes from.

Thanks for raising that up, please let me know if your life got better.